### PR TITLE
SmartAI: Possibility to despawn object with SMART_ACTION_FORCE_DESPAWN

### PR DESCRIPTION
**Pull request by robinsch**
As requested in #15020.

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -1017,16 +1017,24 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
 
             for (ObjectList::const_iterator itr = targets->begin(); itr != targets->end(); ++itr)
             {
-                if (!IsCreature(*itr))
+                if (!IsCreature(*itr) || !IsGameObject(*itr))
```
Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: was this code tested ingame and debugged line by line with both a Creature, a GameObject and an object which is not any of these ?

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -1017,16 +1017,24 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+                        ENSURE_AI(SmartAI, target->AI())->SetDespawnTime(e.action.forceDespawn.delay + 1); // Next tick
+                        ENSURE_AI(SmartAI, target->AI())->StartDespawn();
+                    }
+                    else
+                        target->DespawnOrUnsummon(e.action.forceDespawn.delay);
+                }
+                else if (GameObject* goTarget = (*itr)->ToGameObject())
+                {
+                    if (IsSmartGO(goTarget))
+                        goTarget->SetRespawnTime(e.action.forceDespawn.delay + 1);
```
Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: is there a limit not allowing to despawn a targeted GameObject that is not using SmartAI ?

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (Killyana) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: I just made many testes with creatures and objects and no issues until now.

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 2 above)
Comment (Killyana) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: Creatures and objects not using SmartAI can be despawn, actually It's the case for creatures, npc1 using SAI can despawn npc2 which is not using SAI (AIName=null). I think it must be like this, many other actions works like this.

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (joschiwald) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: useless code

Comment (Killyana) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: This line doesn't exist in the original diff: https://github.com/TrinityCore/TrinityCore/issues/15020#issuecomment-118949343

**Comment by jackpoz** [comment]
I see the code has been updated, please reply to https://github.com/TrinityCore/TrinityCore/pull/15022#discussion_r34023609 anyway

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 2 above)
Comment (joschiwald) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: well tested? i don't believe it works

Comment (Killyana) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: It works fine for me.

Comment (robinsch) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: Line can be removed if you want to allow removing GOs without SmartScripts, I'm not coding DB stuff so I don't know if this is a good idea or it is used in any way.

Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: could you please explain how come SetRespawnTime() is called to despawn the gameobject ? what's the ingame behavior with different e.action.forceDespawn.delay values ?

**Comment by Aokromes** [comment]
Any new on this?

**Comment by Killyana** [comment]
I tested it and it works in cases where we cannot despawn a gobs with lootstate.

**Pull request merged**